crates/polars-ops/src/series/ops/interpolation/interpolate_by.rs

```diff
@@ -0,0 +1,327 @@
+        unsafe { *out.get_unchecked_mut(*out_idx as usize) = v };
+    }
+}
+
+fn interpolate_impl_by_sorted<T, F, I>(
+    chunked_arr: &ChunkedArray<T>,
+    by: &ChunkedArray<F>,
+    interpolation_branch: I,
+) -> PolarsResult<ChunkedArray<T>>
+where
+    T: PolarsNumericType,
+    F: PolarsIntegerType,
+    I: Fn(T::Native, T::Native, &[F::Native], &mut Vec<T::Native>),
+{
+    // This implementation differs from pandas as that boundary None's are not removed.
+    // This prevents a lot of errors due to expressions leading to different lengths.
+    if !chunked_arr.has_validity() || chunked_arr.null_count() == chunked_arr.len() {
+        return Ok(chunked_arr.clone());
+    }
+
+    // At the moment, the algorithm assumes no missing values in the `by` column.
+    let by = by.rechunk();
+    let by_values = by.cont_slice().map_err(|_| {
+        polars_err!(
+            InvalidOperation:
+            "null values in `by` column are not yet supported in 'interpolate_by' expression"
+        )
+    })?;
+
+    // We first find the first and last so that we can set the null buffer.
+    let first = chunked_arr.first_non_null().unwrap();
+    let last = chunked_arr.last_non_null().unwrap() + 1;
+
+    // Fill out with first.
+    let mut out = Vec::with_capacity(chunked_arr.len());
+    let mut iter = chunked_arr.into_iter().enumerate().skip(first);
+    for _ in 0..first {
+        out.push(Zero::zero())
+    }
+
+    let mut low_val = None;
+    let mut low_idx = None;
+    loop {
+        let next = iter.next();
+        match next {
+            Some((idx, Some(v))) => {
+                out.push(v);
+                low_val = Some(v);
+                low_idx = Some(idx);
+            },
+            Some((_idx, None)) => {
+                loop {
+                    match iter.next() {
+                        None => break,         // End of iterator, break.
+                        Some((_, None)) => {}, // Another null.
+                        Some((high_idx, Some(high))) => {
+                            interpolation_branch(
+                                low_val.expect("We started iterating at `first`"),
+                                high,
+                                &by_values[low_idx.unwrap()..=high_idx],
+                                &mut out,
+                            );
+                            out.push(high);
+                            low_val = Some(high);
+                            low_idx = Some(high_idx);
+                            break;
+                        },
+                    }
+                }
+            },
+            None => {
+                break;
+            },
+        }
+    }
+    if first != 0 || last != chunked_arr.len() {
+        let mut validity = MutableBitmap::with_capacity(chunked_arr.len());
+        validity.extend_constant(chunked_arr.len(), true);
+
+        for i in 0..first {
```

Comment:
Maybe do these without bound checks if we can have cases where `first` is large.